#! /bin/sh /usr/share/dpatch/dpatch-run
## 60_fping.dpatch by Niko Tyni <ntyni@iki.fi>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Support the '-S' fping option (#198486)
## DP: Don't try to execute fping if running under the web server.
## DP: (r685 in the upstream repository, modified somewhat)

@DPATCH@
diff -urNad smokeping-2.0.9~/lib/Smokeping/probes/FPing.pm smokeping-2.0.9/lib/Smokeping/probes/FPing.pm
--- smokeping-2.0.9~/lib/Smokeping/probes/FPing.pm	2006-07-07 21:22:14.000000000 +0300
+++ smokeping-2.0.9/lib/Smokeping/probes/FPing.pm	2006-10-25 21:30:18.000000000 +0300
@@ -59,6 +59,8 @@
     	my $binary = join(" ", $self->binary);
 	my $testhost = $self->testhost;
         my $return = `$binary -C 1 $testhost 2>&1`;
+        $self->{enable}{S} = (`$binary -h 2>&1` =~ /\s-S\s/);
+	carp  "NOTE: your fping binary doesn't support source address setting (-S), disabling it" if !$self->{enable}{S};
         croak "ERROR: fping ('$binary -C 1 $testhost') could not be run: $return"
             if $return =~ m/not found/;
         croak "ERROR: FPing must be installed setuid root or it will not work\n" 
@@ -103,10 +105,12 @@
     # pinging nothing is pointless
     return unless @{$self->addresses};
     my @params = () ;
-    push @params , "-b$self->{properties}{packetsize}" if $self->{properties}{packetsize};
+    push @params, "-b$self->{properties}{packetsize}" if $self->{properties}{packetsize};
     push @params, "-t" . int(1000 * $self->{properties}{timeout}) if $self->{properties}{timeout};
     push @params, "-i" . int(1000 * $self->{properties}{mininterval});
     push @params, "-p" . int(1000 * $self->{properties}{hostinterval}) if $self->{properties}{hostinterval};
+    push @params, "-S$self->{properties}{sourceaddress}" if $self->{properties}{sourceaddress} and $self->{enable}{S};
+            
     my @cmd = (
                     $self->binary,
                     '-C', $self->pings, '-q','-B1','-r1',
@@ -139,7 +143,8 @@
 		binary => {
 			_sub => sub {
 				my ($val) = @_;
-        			return "ERROR: FPing 'binary' does not point to an executable"
+        			return undef if $ENV{SERVER_SOFTWARE}; # don't check for fping presence in cgi mode
+				return "ERROR: FPing 'binary' does not point to an executable"
             				unless -f $val and -x _;
 				return undef;
 			},
@@ -190,6 +195,15 @@
 The minimum amount of time between sending a ping packet to any target.
 DOC
 		},
+		sourceaddress => {
+			_re => '\d+\.\d+\.\d+\.\d+',
+			_example => '192.168.0.1',
+			_doc => <<DOC,
+The fping "-S" parameter . From fping(1):
+
+Set source address.
+DOC
+		},
 	});
 }
 
diff -urNad smokeping-2.0.9~/lib/Smokeping/probes/FPing6.pm smokeping-2.0.9/lib/Smokeping/probes/FPing6.pm
--- smokeping-2.0.9~/lib/Smokeping/probes/FPing6.pm	2005-02-22 23:48:07.000000000 +0200
+++ smokeping-2.0.9/lib/Smokeping/probes/FPing6.pm	2006-10-25 21:29:47.000000000 +0300
@@ -46,6 +46,8 @@
       my $self = shift;
       my $h = $self->SUPER::probevars;
       $h->{binary}{_example} = "/usr/bin/fping6";
+      $h->{sourceaddress}{_example} = '::1';
+      $h->{sourceaddress}{_re} = '[0-9A-Fa-f:.]+';
       return $h;
 }
 
